### ISSUE TITLE
[HOLD - new chat component][$1000] Long messages disappear when scrolling through chat history

### ISSUE THREAD

**Issue by Julesssss** [Expensify/App#18067]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Locate a chat with many long messages
2. Scroll up for a while
3. More and more large blank spaces begin to appear

## Expected Result:
- Chats should be visible, no matter how much you scroll back

## Actual Result:
- After scrolling back for a while, chats are not visible

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

Yes, use mobile

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [x] Android / Chrome
- [ ] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** `v1.3.7-0`
**Reproducible in staging?:** Yes
**Reproducible in production?:** Yes
**Issue reported by:** me
**Slack conversation:** [link](https://expensify.slack.com/archives/C049HHMV9SM/p1682584574454669)

https://user-images.githubusercontent.com/10736861/234807831-13681904-4fed-4c17-a724-2487c32c1839.mp4

https://user-images.githubusercontent.com/10736861/234807846-41d59a4f-3659-4d47-8bed-2df7ec66095d.mov

https://user-images.githubusercontent.com/10736861/234807899-fabfe654-6df6-471e-94bc-ec9392946950.mov

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01dc99163ab590ca90</li>
        <li>Upwork Job ID: 1651521261601800192</li>
        <li>Last Price Increase: 2023-04-27</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @jliexpensify (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by jliexpensify** [comment]
Damn, I can repro this. It's very evident with Chronos, but takes more effort with actual DM's.

@Julesssss did you want someone external to take this, and you act as the Supervising Engineer?

**Comment by Julesssss** [comment]
Yeah sounds good. I've only ever noticed this when scrolling through
Chronos so we'll have to get contributors to create lots of long messages

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01dc99163ab590ca90

**Comment by MelvinBot** [comment]
Current assignee @Julesssss is eligible for the External assigner, not assigning anyone new.

**Comment by chenxidev1129** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Long messages disappear when scrolling through chat history
### What is the root cause of that problem?
If the application is not optimized to handle a large number of messages or if it is limited by the amount of memory available on the device, it may struggle to display all the messages properly. This can result in the appearance of large blank spaces or even the disappearance of some messages altogether.

To prevent this issue, chat applications should be designed to efficiently handle large amounts of chat history data. They should also be optimized to make the most of the available resources on the device, such as memory and processing power. Additionally, the application should be regularly updated and maintained to fix any bugs or performance issues that may arise.
### What changes do you think we should make in order to solve the problem?

Improve the loading and display of chat history: The chat application could be optimized to more efficiently load and display chat history, particularly for long messages. This could involve implementing a more efficient algorithm for loading and rendering messages or using pagination to limit the amount of data loaded at once.

Increase the amount of memory available: If the issue is related to insufficient memory on the device, the chat application could be optimized to make the most of the available memory. This could involve implementing caching mechanisms or using compression techniques to reduce the amount of memory needed to store messages.

Implement better error handling: The chat application could be updated to handle errors more gracefully, such as when messages fail to load or display properly. This could involve providing users with clear error messages or allowing them to retry loading messages if they fail to load.

Regularly update and maintain the application: The chat application should be regularly updated and maintained to address any bugs or performance issues that may arise. This could involve implementing new features, fixing bugs, or optimizing the application for new devices or operating systems.
### What alternative solutions did you explore? (Optional)

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by MelvinBot** [comment]
📣 @chenxidev1129! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/231263719-a8a3b7d3-36a0-4d7f-8ad4-cebe1f560fa4.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by chenxidev1129** [comment]
Contributor details
Your Expensify account email: chenxidev1129@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~015d50d1ac51670370

**Comment by MelvinBot** [comment]
✅ Contributor details stored successfully. Thank you for contributing to Expensify!

**Comment by Marlboro11526** [comment]
Contributor details
Your Expensify account email: midastouch12131@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01b551a83e11e05d6a

**Comment by MelvinBot** [comment]
✅ Contributor details stored successfully. Thank you for contributing to Expensify!

**Comment by sobitneupane** [comment]
@chenxidev1129 Thanks for [your proposal](https://github.com/Expensify/App/issues/18067#issuecomment-1525341915).

Looks like you are a new contributor. Please go through [the contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) if you have not already. 
We expect some concrete proposal. Your proposal is quite generic, please try to be specific. Also please make use of permalink to point where the issue exists in app whenever possible. You can go through some other issues to know about the expectations in a proposal.

**Comment by chenxidev1129** [comment]
Okay. Thanks. I will check again @sobitneupane

**Comment by chafikchaban** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
React Native Flatlist Performance issue. Long messages are randomly not rendered in a large messages list
### What is the root cause of that problem?

`InvertedFlatList` used in `ReportActionsList` is not well configured:
-  the `initialNumToRender` which we calculate to return the maximum number of smallest space taking items that could fit the available visible space ( which imo is a little random ) 

According to the [React Native docs: initialnumtorender](https://reactnative.dev/docs/0.67/optimizing-flatlist-configuration#initialnumtorender):

 "_Setting a low initialNumToRender may cause blank areas, especially if it's too small to cover the viewport on initial render._"

-  that is considering `BaseInvertedFlatList`'s default configuration. more specifically `maxtorenderperbatch` which is set to `1`. which was set in [this commit](https://github.com/Expensify/App/commit/54ecd70161d7f42755a703f85c69796c3e710d65) by @marcaaron in order to improve overall performance

According to the [React Native docs: maxToRenderPerBatch](https://reactnative.dev/docs/0.67/optimizing-flatlist-configuration#maxtorenderperbatch):

 "_Setting a bigger number means less visual blank areas when scrolling (increases the fill rate)._"


### What changes do you think we should make in order to solve the problem?
we need to tweak these values. the best way to do it is:
- setting `initialnumtorender` to the length of the array ( this does actually affects the list's initial rendering performance just a little bit, but is optimal for scrolling and guarantees a smooth experience with no jumping on scroll )

### What alternative solutions did you explore? (Optional)
- using a higher value of `maxToRenderPerBatch` ( even though just setting it to `2` solves our issue, the jumping of next batch being rendered is actualy visible and annoying tbh)

**Comment by romulo114** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When scrolling chat history, messages sometimes disappear.

### What is the root cause of that problem?
We're using `FlatList` component of react-native for displaying messages history. This component uses `getItemLayout` props to calculate the item's offset and height. In our project, We calculate the offset of the item using the following code.

https://github.com/Expensify/App/blob/a7863493a03970cdc0687621ec40dc75aea99477/src/components/InvertedFlatList/BaseInvertedFlatList.js#L57-L112

This algorithm assumes that the `measureItemLayout` is called in order, but this is not the case. Actually, `measureItemLayout` callback is called in different order like belows. This is the main reason.

![image](https://user-images.githubusercontent.com/117526965/235340230-efa8afbf-5990-42fa-859a-e2fcd2dfce16.png)

### What changes do you think we should make in order to solve the problem?

In order to solve this issue, we have to update the `getItemLayout` and `measureItemLayout` method so that get the offset of the item correctly. I was able to fix the issue successfully by assuming that `measureItemLayout` is called in any order.

<details>
    <summary>Result</summary>


https://user-images.githubusercontent.com/117526965/235340567-9263b66f-2eb8-41e3-901e-602e12633d15.mp4


</details>

### What alternative solutions did you explore? (Optional)
NA

**Comment by Julesssss** [comment]
Thanks all! All are well explained and make sense. 

@romulo114 I'm curious what your solution would look like?

**Comment by romulo114** [comment]
@Julesssss 
Please take a look.

https://github.com/Expensify/App/blob/a7863493a03970cdc0687621ec40dc75aea99477/src/components/InvertedFlatList/BaseInvertedFlatList.js#L93-L112

We can get the item's height correctly but the offset isn't correct. As you can see, we calculate the offset of the item based on previous offset, which is assuming that this callback is called in order. Actually this callback isn't called in order.

In order to solve this issue, I get the item's length(height) only in this callback(by removing the [line 110](https://github.com/Expensify/App/blob/a7863493a03970cdc0687621ec40dc75aea99477/src/components/InvertedFlatList/BaseInvertedFlatList.js#L110), and get the offset of the item when `getItemLayout` is called.

In `getItemLayout` function, I get the offset by cumulating the heights previous items and put them in `sizeMap` when offset isn't available. Of course, offset is available, I get the offset from sizeMap. If height of an item isn't available yet, I use `initialRowHeight` prop of the component.

**Comment by bernhardoj** [comment]
Based on this comment from Pagination tracking https://github.com/Expensify/App/issues/12054#issuecomment-1518207853, `getItemLayout` could be removed.

**Comment by romulo114** [comment]
> Based on this comment from Pagination tracking [#12054 (comment)](https://github.com/Expensify/App/issues/12054#issuecomment-1518207853), `getItemLayout` could be removed.

As you can see [here](https://reactnative.dev/docs/flatlist#getitemlayout), it is more performant to use when there is a lot of messages. Can you specify the clear reason that getItemLayout should be removed.

**Comment by Julesssss** [comment]
Thanks @romulo114! @sobitneupane please review [this proposal](https://github.com/Expensify/App/issues/18067#issuecomment-1528956572) and let me know what you think.

**Comment by Julesssss** [comment]
If we like the proposal it's probably fine to not see the whole diff, but if you do have it written already please share.

**Comment by chafikchaban** [comment]
If we're going that route, i think the biggest challenge is to find a way to make sure that the measuring occurs first, otherwise we're just assuming and estimating item sizes anyway. which to my understanding is a big no-no to pass to the Flatlist's `getItemLayout` 

also: the order in which `measureItemLayout` is called doesn't  really matter since we store the measurements in the **sizeMap** accordingly to the measured item's index on the initial list. 
 
**Concern:**  order of execution between the Flatlist's `getItemLayout` and the item's `onLayout`

---

But also digging deeper, does it even make sense in our case to use `getItemLayout` ? 
According to the docs [here](https://reactnative.dev/docs/optimizing-flatlist-configuration#use-getitemlayout) : 

"`getItemLayout` should only be used when _**all your list item components have the same height (or width, for a horizontal list)**_"

and [here](https://reactnative.dev/docs/flatlist#getitemlayout):

"`getItemLayout` is an optional optimisation that allows skipping the measurement of dynamic content **if you know the size (height or width) of items ahead of time**. getItemLayout is efficient **if you have fixed size items**_"

which is not the case here 🤔 

@Julesssss  @sobitneupane  thoughts ?

**Comment by romulo114** [comment]
> But also digging deeper, does it even make sense in our case to use `getItemLayout` ? According to the docs [here](https://reactnative.dev/docs/optimizing-flatlist-configuration#use-getitemlayout) :
> 
> "`getItemLayout` should only be used when _**all your list item components have the same height (or width, for a horizontal list)**_"
> 
> and [here](https://reactnative.dev/docs/flatlist#getitemlayout):
> 
> "`getItemLayout` is an optional optimisation that allows skipping the measurement of dynamic content **if you know the size (height or width) of items ahead of time**. getItemLayout is efficient **if you have fixed size items**_"


@chafikchaban 

They are saying `getItemLayout` can be used to optimize the performance especially when items have the same height. This doesn't mean we should not use this when heights are different. For example, `getItemLayout` would work great when we can get the items' heights easily.

> also: the order in which `measureItemLayout` is called doesn't really matter since we store the measurements in the **sizeMap** accordingly to the measured item's index on the initial list.

If you took a look at the source, you can notice that the order is important in calculating offsets.

**Comment by jliexpensify** [comment]
Just giving this a quick bump @Julesssss and @sobitneupane - I tested this again with v1.3.9-19 and it's still happening, albeit not as bad as it was before. Would any of the above proposals work?

**Comment by romulo114** [comment]
@jliexpensify @Julesssss @sobitneupane 
Here are my fixes. They work perfect.

![image](https://user-images.githubusercontent.com/117526965/236131258-f54a4986-bdce-4a52-af14-bd1e34f1f3e6.png)

![image](https://user-images.githubusercontent.com/117526965/236133720-653bc734-e919-4d22-946b-89428e86be80.png)

**Comment by sobitneupane** [comment]
> If we're going that route, i think the biggest challenge is to find a way to make sure that the measuring occurs first, otherwise we're just assuming and estimating item sizes anyway. which to my understanding is a big no-no to pass to the Flatlist's getItemLayout
Concern: order of execution between the Flatlist's getItemLayout and the item's onLayout

@chafikchaban has raised some valid concern [here](https://github.com/Expensify/App/issues/18067#issuecomment-1531763903). What do you think about it @romulo114?


The other thing is [the solution](https://github.com/Expensify/App/issues/18067#issuecomment-1534193812) looks computationally very expensive.  I think it will worsen the performance of big chats.

**Comment by Julesssss** [comment]
Thanks, @sobitneupane. I'm thinking that we should take this chance to ask one of our external partner teams to dive deeper into this problem.

**Comment by romulo114** [comment]
> If we're going that route, i think the biggest challenge is to find a way to make sure that the measuring occurs first, otherwise we're just assuming and estimating item sizes anyway. which to my understanding is a big no-no to pass to the Flatlist's getItemLayout
> Concern: order of execution between the Flatlist's getItemLayout and the item's onLayout
> 
> @chafikchaban has raised some valid concern [here](https://github.com/Expensify/App/issues/18067#issuecomment-1531763903). What do you think about it @romulo114?

As you can see [here](https://stackoverflow.com/a/48682288), React Native `FlatList` optimises list view performance by only rendering rows that are currently visible on the screen, and unmounting rows that have been scrolled past.

If we don't know the item's height in advance, we can't determines the total height of the scroll view. That's why FlatList uses `getItemLayout` **_for getting the visible elements_**(so needs to be rendered), not for getting the exact offset of the item and total height of the scroll view.

We are setting item height as the _minimum item height_(32 in our case) as you can see below. This ensures visible items always to be rendered.

https://github.com/Expensify/App/blob/75d03a479ae1e98eca61d631e24380904f470b3b/src/components/InvertedFlatList/BaseInvertedFlatList.js#L16-L17

https://github.com/Expensify/App/blob/75d03a479ae1e98eca61d631e24380904f470b3b/src/pages/home/report/ReportActionsList.js#L153

Once items are rendered, we can get their offsets and heights. As we scroll the view, FlatList calculates their offset based on these values and determines the visible elements correctly. 

> The other thing is [the solution](https://github.com/Expensify/App/issues/18067#issuecomment-1534193812) looks computationally very expensive. I think it will worsen the performance of big chats.

Please take a look at the source carefully. Its calculation complexity is O(1). Only a few math operations for each item. And we cache all items' offsets and heights to avoid re-calculation. Performance isn't a problem at all, I think.

**Comment by OlimpiaZurek** [comment]
Hi, I'm Olimpia from Callstack - expert contributor group - I'll be happy to work on this issue.

**Comment by Julesssss** [comment]
Hey, thanks @OlimpiaZurek! Please let me know if you have any questions, here on in Slack

**Comment by OlimpiaZurek** [comment]
Hey @Julesssss!. I checked possible solutions to this issue and found that replacing `FlatList` with `FlashList` seems to be the best and works perfectly.

FlashList
![FlashList](https://user-images.githubusercontent.com/19835085/236792682-38f58a3f-49b1-4ac1-952a-fc096d059e45.gif)

FlatList
![FlatList](https://user-images.githubusercontent.com/19835085/236793516-0c13c4ed-4767-4212-b28b-101b41a2207e.gif)


But as far as I know you are working on building a custom list component which probably solves this problem as well. I don't know what stage the work on this new component is at, but I think in this situation moving to `FlashList` until the new list component is implemented seems pointless. 

Maybe the best option is to hold off on this issue until the new list component is ready and then check if the issue still occurs?

**Comment by romulo114** [comment]
@OlimpiaZurek 
Thank you for your explanation.
Will you explain more about FlashList?

**Comment by jliexpensify** [comment]
No overdue, waiting on @Julesssss input to see if we want this on HOLD.

**Comment by Julesssss** [comment]
Thanks, @OlimpiaZurek, that's very useful. I'm curious if FlatList is using Android RecyclerList natively, as the pattern sounds very similar (recycling the view instead of recreating).

**Comment by Julesssss** [comment]
Also, I'm not familiar with the internal list we're working on. I thought we were planning to migrate to FlatList for some reason.

**Comment by OlimpiaZurek** [comment]
> Thanks, @OlimpiaZurek, that's very useful. I'm curious if FlatList is using Android RecyclerList natively, as the pattern sounds very similar (recycling the view instead of recreating).

@Julesssss Do you mean `FlashList` ? It uses [Recyclerlistview](https://github.com/Flipkart/recyclerlistview) to leverage its recycling capability.



> Also, I'm not familiar with the internal list we're working on. I thought we were planning to migrate to FlatList for some reason.

Take a look at these comments from other issues:
https://github.com/Expensify/App/issues/17209#issuecomment-1532051438
https://github.com/Expensify/App/issues/16935#issuecomment-1531867407

**Comment by jliexpensify** [comment]
Hi @Julesssss - just to confirm: this issue should be placed on HOLD?

**Comment by Julesssss** [comment]
Yeah 👍

**Comment by jliexpensify** [comment]
Still on HOLD

**Comment by Julesssss** [comment]
Still held, will be for a while

**Comment by jliexpensify** [comment]
On hold

**Comment by jliexpensify** [comment]
On hold

**Comment by jliexpensify** [comment]
On hold

**Comment by Julesssss** [comment]
Hey @OlimpiaZurek, just wondering if this still needs to be on hold?

**Comment by OlimpiaZurek** [comment]
As far as I know, we are rolling out `Flashlist,` so I think this should be put on hold until the implementation is complete

**Comment by Julesssss** [comment]
Ah thanks. @jasperhuangg lets just close this? Is the original issue even still reproducable, I don't think I have seen it in a while

**Comment by Julesssss** [comment]
Still on hold

**Comment by Julesssss** [comment]
Hey @jliexpensify, I kinda think we should close this. It seems unlikely that we'll be prioritising the list component changes, and both the linked issues have been closed.

**Comment by jliexpensify** [comment]
No worries, closing! Nothing to payout here.
